src/sgame/sg_buildable.cpp

```diff
@@ -2291,80 +2292,61 @@
 
 void G_LayoutLoad()
 {
-	fileHandle_t f;
-	int          len;
-	char         *layout, *layoutHead;
-	char         map[ MAX_QPATH ];
-	char         buildName[ MAX_TOKEN_CHARS ];
-	int          buildable;
-	vec3_t       origin = { 0.0f, 0.0f, 0.0f };
-	vec3_t       angles = { 0.0f, 0.0f, 0.0f };
-	vec3_t       origin2 = { 0.0f, 0.0f, 0.0f };
-	vec3_t       angles2 = { 0.0f, 0.0f, 0.0f };
-	char         line[ MAX_STRING_CHARS ];
-	int          i = 0;
-	const buildableAttributes_t *attr;
-
 	if ( !level.layout[ 0 ] || !Q_stricmp( level.layout, S_BUILTIN_LAYOUT ) )
 	{
 		return;
 	}
 
-	trap_Cvar_VariableStringBuffer( "mapname", map, sizeof( map ) );
-	len = BG_FOpenGameOrPakPath( va( "layouts/%s/%s.dat", map, level.layout ), f );
-
+	int f;
+	std::string path = Str::Format( "layouts/%s/%s.dat" , Cvar::GetValue( "mapname" ), level.layout );
+	int len = BG_FOpenGameOrPakPath( path, f );
 	if ( len < 0 )
 	{
 		Log::Warn( "layout %s could not be opened", level.layout );
 		return;
 	}
 
-	layoutHead = layout = (char*) BG_Alloc( len + 1 );
-	trap_FS_Read( layout, len, f );
-	layout[ len ] = '\0';
+	std::string data;
+	data.resize( len );
+	trap_FS_Read( &data[ 0 ], len, f );
 	trap_FS_FCloseFile( f );
 
-	while ( *layout )
+	for ( Parse_WordListSplitter parse{std::move( data ), "\n"}; *parse; ++parse )
 	{
-		if ( i >= (int) sizeof( line ) - 1 )
-		{
-			Log::Warn( "line overflow in %s before \"%s\"",
-			          va( "layouts/%s/%s.dat", map, level.layout ), line );
-			break;
-		}
+		char         buildName[ 128 ];
+		vec3_t       origin;
+		vec3_t       angles;
+		vec3_t       origin2;
+		vec3_t       angles2;
 
-		line[ i++ ] = *layout;
-		line[ i ] = '\0';
+		int n = sscanf( *parse, "%127s %f %f %f %f %f %f %f %f %f %f %f %f",
```

Comment:
Can you elaborate on that

Reply:
That's a good catch. The intention is to skip the warning if the line is just empty. It appeared to work in my tests, but that was because a completely empty line is just filtered out by the WordListSplitter. The EOF case may happen if the line contains some whitespace.

Reply:
But did that not check if the result is not 13? EOF can never be 13

Reply:
The old code was
```
if ( n == 0 )
{
	Log::Warn( "bad line in layout %s", path );
}
```
With the intention to check whether the line was blank. If the line is blank it will have a negative EOF value, rather than 0. sscanf has a weird behavior where it reacts differently if the end of the string is hit before parsing the first specifier vs. after.